enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,247 @@
+
+Users can also be added to groups in the IdP, in which case they are more
+likely to willingly cause a resync by `oc login` themselves.
+
+## Design Details
+
+### Open Questions
+
+### Test Plan
+
+The authentication operator already has a way to configure Keycloak and
+communicate with its API. Use that to add test users and groups to test:
+
+1. user can still log in when groups are configured but no group
+   claims are received
+2. on login, user gets added to a previously non-existent group when
+   the group appears in the configured claims
+3. on login, user gets added to an existing group that appears in the
+   retrieved group claims
+4. on login, user gets removed from any existing group that was created
+   for the given identity provider (per annotation) when the group
+   no longer appears among those retrieved from the OIDC
+
+
+### Follow-up Work
+
+To synchronize groups in a time-based manner, an additional controller in the
+oauth-server would be necessary. OIDC does not provide a standardized group
+membership and reverse group membership lookups. That means that such a controller
+would have to cache the access tokens (and possibly refresh tokens) retrieved
+from the OIDC after a successful user login, and use them to query the UserInfo
+endpoint. This could be troublesome, however, because:
+
+1. the access token we get from the OIDC might be shortlived, if there are no
+   refresh tokens we can't synchronize the groups for the given user for too long
+2. we might end up generating a lot of traffic to the OIDC provider
+3. the OIDC may not expose/implement the UserInfo endpoint
+
+### Graduation Criteria
+
+The feature should be considered mature once it lands.
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
+
+There are two drawbacks to the current solution:
+
+1. the groups go out of sync between logins
+- This is quite a common problem with all systems that receive delegated identities.
+  Third party vendors don't usually provide a standardized watch channel that could be
+  effectively and reliably observed, and so most systems either only synchronize on user
+  login or attempt to perform group retrievals in regular intervals, but real real-time
+  synchronization is usually impossible.
+
+2. the groups are assigned to a user, not an `<identity provider:user>` touple
+- Some configurations allow a user to be sourced by multiple identity providers, where
+  each of the providers might have different groups configured. The implemenentation
+  described in this enhancement does not distinguish the groups when the user attempts
+  to use their token and just uses all the OpenShift groups the user appears in.
+
+## Alternatives
```

Comment:
Can you describe the shortcomings of using OIDC directly from the kube-apiserver?  I suspect that write up may become even more important over time.

I think the primary issue is the usage of serviceaccounts and their tokens as automatic oauthclients which can be used for in-cluster addons automatically on openshift.  This shortcoming may be enough to justify work like this, but it may also justify why certain integration paths for external token issuers fall short and give a concrete path for another project to become a token issuer.

Reply:
That's a good point, I'll try to think about this and will add a section here.

Reply:
Outlined a few issues that came to my mind